src/history/hs_rec.c

```diff
@@ -911,25 +930,103 @@ __hs_delete_reinsert_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, ui
       session, ts == 1 || ts == WT_TS_NONE, "out-of-order timestamp update detected");
+    return (ret);
+}
+
+/*
+ * __hs_insert_pos_counter --
+ *     Get the count for the insert position for zero timestamped records.
+ */
+static int
+__hs_insert_pos_counter(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, uint32_t btree_id,
+  const WT_ITEM *key, uint64_t *counter)
+{
+    WT_DECL_RET;
+    WT_ITEM hs_key, hs_value;
+    wt_timestamp_t hs_start_ts;
+    uint64_t hs_counter;
+    uint32_t hs_btree_id;
+#ifdef HAVE_DIAGNOSTIC
+    int cmp;
+#endif
+    WT_CLEAR(hs_key);
+    WT_CLEAR(hs_value);
+    WT_UNUSED(btree_id);
+#ifndef HAVE_DIAGNOSTIC
+    WT_UNUSED(key);
+#endif
+
+    for (; ret == 0; ret = hs_cursor->next(hs_cursor)) {
+        /* We shouldn't have crossed the btree and user key search space. */
+        WT_RET(hs_cursor->get_key(hs_cursor, &hs_btree_id, &hs_key, &hs_start_ts, &hs_counter));
+        WT_ASSERT(session, hs_btree_id == btree_id);
+#ifdef HAVE_DIAGNOSTIC
+        WT_RET(__wt_compare(session, NULL, &hs_key, key, &cmp));
+        WT_ASSERT(session, cmp == 0);
+#endif
+        ++(*counter);
+    }
+    if (ret == WT_NOTFOUND)
+        ret = 0;
+    return (ret);
+}
+
+/*
+ * __hs_delete_reinsert_from_pos --
+ *     Delete updates in the history store if the start timestamp of the update is larger or equal
+ *     to the specified timestamp and optionally reinsert them with ts-1 timestamp.
+ */
+static int
+__hs_delete_reinsert_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, uint32_t btree_id,
+  const WT_ITEM *key, wt_timestamp_t ts, bool reinsert, bool no_ts_tombstone, bool non_ts_updates,
+  bool error_on_ts_ordering, uint64_t *counter, WT_TIME_WINDOW *upd_tw)
+{
+    WT_CURSOR *hs_insert_cursor;
+    WT_CURSOR_BTREE *hs_cbt;
+    WT_DECL_RET;
+    WT_ITEM hs_key, hs_value;
+    WT_TIME_WINDOW hs_insert_tw, *twp;
+    wt_timestamp_t hs_durable_start_ts, hs_durable_stop_ts, hs_start_ts;
+    uint64_t cache_hs_order_lose_durable_timestamp, cache_hs_order_reinsert, cache_hs_order_remove;
+    uint64_t hs_counter, hs_upd_type;
+    uint32_t hs_btree_id;
+#ifdef HAVE_DIAGNOSTIC
+    int cmp;
+#endif
+    char ts_string[5][WT_TS_INT_STRING_SIZE];
+
+    hs_insert_cursor = NULL;
+    hs_cbt = __wt_curhs_get_cbt(hs_cursor);
+    WT_CLEAR(hs_key);
+    WT_CLEAR(hs_value);
+    WT_UNUSED(upd_tw);
+    WT_UNUSED(error_on_ts_ordering);
+
+    cache_hs_order_lose_durable_timestamp = cache_hs_order_reinsert = cache_hs_order_remove = 0;
+#ifndef HAVE_DIAGNOSTIC
+    WT_UNUSED(key);
+#endif
+
     /*
      * Fail the eviction if we detect any timestamp ordering issue and the error flag is set. We
      * cannot modify the history store to fix the update's timestamps as it may make the history
-     * store checkpoint inconsistent.
+     * store checkpoint inconsistent. We are not reinserting for non-timestamped updates, continue.
      */
-    if (error_on_ts_ordering) {
+    if (error_on_ts_ordering && !non_ts_updates) {
```

Comment:
If we are not reinserting for non_ts_updates, why the check is !non_ts_updates? Also why we need to return EBUSY? In this case, should we return an error or doing no-op?

Reply:
It is still possible for the scenario that we have `non_ts_update` is true, and `reinsert` is false (this is in the case we want to have `non_ts_updates` for deletes) we still enter this function for removal but do not want to return 
 `EBUSY` here.  Otherwise, for the normal scenario where `non_ts_update` is true (`reinsert` is also set to true - but we wish to skip) we do not enter this function at all so it is a no-op.

Reply:
Then the comment change is misleading. Can you explain this in the comment?

Reply:
Updated the comment here [b6900cf](https://github.com/wiredtiger/wiredtiger/pull/10686/commits/b6900cf433be480879bbfe2be8d51f4507fcea13)